### DISCUSSION TITLE
Live filter files picker to subdirectory?

### DISCUSSION THREAD

**Discussion by heygarrett** [ibhagwan/fzf-lua#1402] [Q&A] [answered]
I've customized the files picker to include directories in the search results. Is there a way to to set up an action that updates the files picker to search within the selected subdirectory?

This _sort of_ works:

```lua
["ctrl-f"] = function(selected)
	local relative_path = require("fzf-lua.path").entry_to_file(selected[1]).stripped
	vim.api.nvim_input("<esc>")
	require("fzf-lua").files({ cwd = vim.fs.joinpath(vim.env.PWD, relative_path) })
end,
```

But I end up in normal mode and this error is displayed:

```
ERROR
[Fzf-lua] fn_selected threw an error: ...rett/.local/share/nvim/lazy/fzf-lua/lua/fzf-lua/core.lua:243: vim/_editor.lua:0: nvim_exec2()..FileType Autocommands for "*": Vim(append):Error executing lua callback: /Users/garrett/.confi
g/nvim/lua/my/settings/options.lua:38: Vim:Can't re-enter normal mode from terminal mode
stack traceback:
        [C]: in function 'normal'
        /Users/garrett/.config/nvim/lua/my/settings/options.lua:38: in function </Users/garrett/.config/nvim/lua/my/settings/options.lua:31>
        [C]: in function 'resume'
        ...rrett/.local/share/nvim/lazy/fzf-lua/lua/fzf-lua/fzf.lua:323: in function <...rrett/.local/share/nvim/lazy/fzf-lua/lua/fzf-lua/fzf.lua:304>
stack traceback:
        ...rett/.local/share/nvim/lazy/fzf-lua/lua/fzf-lua/core.lua:240: in function <...rett/.local/share/nvim/lazy/fzf-lua/lua/fzf-lua/core.lua:234>
        [builtin#36]: at 0x0100eb0760
        ...rett/.local/share/nvim/lazy/fzf-lua/lua/fzf-lua/core.lua:243: in function 'files'
        /Users/garrett/.config/nvim/lua/my/plugins/fzf-lua.lua:164: in function 'action'
        ...t/.local/share/nvim/lazy/fzf-lua/lua/fzf-lua/actions.lua:98: in function 'act'
        ...rett/.local/share/nvim/lazy/fzf-lua/lua/fzf-lua/core.lua:148: in function 'fn_selected'
        ...rett/.local/share/nvim/lazy/fzf-lua/lua/fzf-lua/core.lua:233: in function <...rett/.local/share/nvim/lazy/fzf-lua/lua/fzf-lua/core.lua:232>
        [C]: in function 'xpcall'
        ...rett/.local/share/nvim/lazy/fzf-lua/lua/fzf-lua/core.lua:232: in function <...rett/.local/share/nvim/lazy/fzf-lua/lua/fzf-lua/core.lua:224>
```

The error refers to an auto-command I have configured in a different file (for some reason):
```lua
vim.api.nvim_create_autocmd("FileType", {
	desc = "restore cursor position",
	group = group,
	callback = function()
		local exclude = { "diff", "gitcommit", "gitrebase" }
		if vim.tbl_contains(exclude, vim.bo.filetype) then
			return
		end
		local position_line = vim.api.nvim_buf_get_mark(0, [["]])[1]
		if position_line >= 1 and position_line <= vim.api.nvim_buf_line_count(0) then
			vim.cmd.normal({ -- line 38
				bang = true,
				args = { [[g`"]] },
			})
		end
	end,
})
```

**Comment by ibhagwan** [comment, marked as answer]
> The error refers to an auto-command I have configured in a different file (for some reason):

Fzf-Lua sets the file type to fzf so it also triggers the auto command, you should probably add it to the exclude list.

Another option is to reuse the existing fzf-lua windows by wrapping the action in `{}`:

```lua
["ctrl-f"] = { function(selected)
	local relative_path = require("fzf-lua.path").entry_to_file(selected[1]).stripped
        -- this line won’t be needed
	-- vim.api.nvim_input("<esc>")
	require("fzf-lua").files({ cwd = vim.fs.joinpath(vim.env.PWD, relative_path) })
end },
```

**Reply by heygarrett** [reply]
Wrapping the function in a table doesn't seem to make a difference, but excluding "fzf" from the auto-command does get rid of the error.

Is there a better way to do this? What I'm doing feels like a workaround rather than a solution.

**Reply by ibhagwan** [reply]
Yes, it’s a workaround for a bug your auto command creates :)

You’re asking neovim to preform something on setting file types rather indiscriminately, neovim does what you asked it to do which results in a bug, this is not a workaround, this is a **fix** for a bug your code created, which is not taking into account terminal buffers which can’t accept `vim.cmd.normal` commands.

**Reply by heygarrett** [reply]
That's fair. 😛 I've updated my auto-command to better reflect the example in the docs (`:help restore-cursor`) which resolves the issue without me needing to explicitly exclude the fzf file type.

<details><summary>Updated auto-command</summary>

```lua
vim.api.nvim_create_autocmd("BufRead", {
	desc = "restore cursor position",
	group = group,
	callback = function(event_args)
		vim.api.nvim_create_autocmd("FileType", {
			group = group,
			buffer = event_args.buf,
			once = true,
			callback = function()
				local exclude = { "diff", "gitcommit", "gitrebase" }
				if vim.tbl_contains(exclude, vim.bo.filetype) then
					return
				end
				local position_line = vim.api.nvim_buf_get_mark(0, [["]])[1]
				if
					position_line >= 1
					and position_line <= vim.api.nvim_buf_line_count(0)
				then
					vim.cmd.normal({
						bang = true,
						args = { [[g`"]] },
					})
				end
			end,
		})
	end,
})
```

</details>

**Reply by heygarrett** [reply]
@ibhagwan One additional question: Is there a better way to get the full path of the selected ~file~ directory?

**Reply by ibhagwan** [reply]
What you did is acceptable but it doesn't check if the paths is already absolute, you can utilize fzf-lua's path module for that:
```lua
  local fzf = require("fzf-lua")
  local path = fzf.path
  local file = path.entry_to_file(selected[1])
  if not path.is_absolute(file.path) then
    file.path = path.join({ vim.uv.cwd(), file.path })
  end
```

**Reply by heygarrett** [reply]
Ah, I hadn't even thought of that. Thanks!

**Reply by heygarrett** [reply]
Another thing I hadn't thought of: this only works once; attempting to filter a second time into another subdirectory doesn't work because it only joins the selected entry text with the cwd. Is there a way to always get the absolute path of the entry? Or would that be a feature request?

**Reply by ibhagwan** [reply]
> Another thing I hadn't thought of: this only works once; attempting to filter a second time into another subdirectory doesn't work because it only joins the selected entry text with the cwd. Is there a way to always get the absolute path of the entry? Or would that be a feature request?

That’s unrelated to a feature request, it’s on you to maintain the state vs display, even if the entry is just the directory and you join `cwd` why wouldn’t it be turned into an absolute path?

**Reply by heygarrett** [reply]
Hm, I assumed fzf-lua would always have absolute paths for the files listed in the picker, but I'm realizing now that it likely doesn't need them most of the time.

This is the problem I'm running into:  
- I have directory structure A/B/C and A is the cwd
- I use this action to filter results to only directory B (eg, `require("fzf-lua").files({ cwd = "A/B" })`)
- I use this action _again_ to try to filter results to only directory C, but what ends up happening is the `cwd` passed to `files()` ends up being "A/C" (the actual cwd joined with the selected entry) but that's not a real path.

I'll experiment and see if I can come up with a decent way to track the "current" cwd, as seen by the files picker.

**Reply by heygarrett** [reply]
Oh, there's an `opts` parameter to the function that tracks the "current" cwd with `opts.cwd`. That works perfectly for what I'm trying to do.

**Reply by ibhagwan** [reply]
Correct, `opts` contains all supplied options including the `cwd` if it was supplied so make sure to use `opts.cwd or vim.uv.cwd()`.
